File: examples/laravel/app/Services/MyOtpClient.php
<?php

namespace App\Services;

use App\Exceptions\MyOtpException;
use Illuminate\Support\Facades\Http;

/**
 * Server-side client for the MyOTP.App REST API.
 *
 * Bound as a singleton in AppServiceProvider; injected into controllers.
 * The API key is read from config('services.myotp.key'), which is populated
 * from MYOTP_API_KEY at config-load time.
 *
 * API reference: https://api.myotp.app
 */
class MyOtpClient
{
    public function __construct(
        private readonly string $apiKey,
        private readonly string $baseUrl,
    ) {
    }

    /**
     * @throws MyOtpException
     */
    public function generateOtp(string $phoneNumber, string $channel = 'sms'): array
    {
        return $this->post('/generate_otp', [
            'phone_number' => $phoneNumber,
            'channel' => $channel,
        ]);
    }

    /**
     * @throws MyOtpException
     */
    public function verifyOtp(string $otp, ?string $messageId = null, ?string $phoneNumber = null): array
    {
        $payload = ['otp' => $otp];
        if ($messageId !== null) {
            $payload['message_id'] = $messageId;
        }
        if ($phoneNumber !== null) {
            $payload['phone_number'] = $phoneNumber;
        }
        return $this->post('/verify_otp', $payload);
    }

    /**
     * @throws MyOtpException
     */
    private function post(string $path, array $payload): array
    {
        if ($this->apiKey === '') {
            throw new MyOtpException(500, 'MYOTP_API_KEY is not configured');
        }

        $response = Http::withHeaders([
            'Content-Type' => 'application/json',
            'X-API-Key' => $this->apiKey,
        ])
            ->timeout(15)
            ->post(rtrim($this->baseUrl, '/') . $path, $payload);

        $data = $response->json() ?? [];

        if (! $response->successful()) {
            $message = $data['message'] ?? $data['error'] ?? 'MyOTP error ' . $response->status();
            throw new MyOtpException($response->status(), (string) $message);
        }

        return $data;
    }
}

File: examples/rails/app/lib/myotp_client.rb
# Server-side client for the MyOTP.App REST API.
#
# The API key is read from Rails configuration (loaded from env at boot in
# config/application.rb) so it never leaks to the browser-bound asset pipeline.
#
# API reference: https://api.myotp.app
require "net/http"
require "uri"
require "json"

module MyotpClient
  Error = Class.new(StandardError) do
    attr_reader :status

    def initialize(status, message)
      super(message)
      @status = status
    end
  end

  module_function

  def generate_otp(phone_number:, channel: "sms")
    post("/generate_otp", phone_number: phone_number, channel: channel)
  end

  def verify_otp(otp:, message_id: nil, phone_number: nil)
    payload = { otp: otp }
    payload[:message_id] = message_id if message_id
    payload[:phone_number] = phone_number if phone_number
    post("/verify_otp", payload)
  end

  def post(path, payload)
    api_key = Rails.configuration.x.myotp.api_key.to_s
    raise Error.new(500, "MYOTP_API_KEY is not configured") if api_key.empty?

    base_url = Rails.configuration.x.myotp.base_url.to_s.chomp("/")
    uri = URI.parse("#{base_url}#{path}")

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == "https")
    http.read_timeout = 15
    http.open_timeout = 10

    request = Net::HTTP::Post.new(uri.request_uri)
    request["Content-Type"] = "application/json"
    request["X-API-Key"] = api_key
    request.body = payload.to_json

    response = http.request(request)
    data = parse_json(response.body)

    if response.is_a?(Net::HTTPSuccess)
      data
    else
      message = data["message"] || data["error"] || "MyOTP error #{response.code}"
      raise Error.new(response.code.to_i, message)
    end
  end

  def parse_json(body)
    return {} if body.nil? || body.empty?
    JSON.parse(body)
  rescue JSON::ParserError
    {}
  end
end
